.github/workflows/tag-build.yaml

```diff
@@ -1,10 +1,220 @@
+                          return
+                      fi
+          
+                      # Case: tag event with prerelease version.
+                      if [ "${GITHUB_EVENT_REF_NAME#*-}" != "$GITHUB_EVENT_REF_NAME" ]; then
+                          echo "environment=build-stage"
+                          echo "version=$GITHUB_EVENT_REF_NAME"
+                          return
+                      fi
+          
+                      # Case: tag event with release version. Only this
+                      # should go to prod.
+                      echo "environment=build-prod"
+                      echo "version=$GITHUB_EVENT_REF_NAME"
+                      ;;
+                  *)
+                      >&2 echo "Unknown GHA event $GITHUB_EVENT_NAME, failing" 
+                      exit 1
+                      ;;
+              esac
+          }
+          
+          get_output_vars >> "$GITHUB_OUTPUT"
+          
+          # Validate the semver
+          . "$GITHUB_OUTPUT"  # Load the variables into the current environment
+          (echo "$version" | grep -qP "$SEMVER_REGEX") || (echo "The artifact version $version is not a valid semver-coerced value"; exit 1)
+        
+          # Log the build details
+          echo "Built config:" | tee -a "$GITHUB_STEP_SUMMARY"
+          cat "$GITHUB_OUTPUT" | while read line; do
+            echo "* $line" | tee -a "$GITHUB_STEP_SUMMARY"
+          done
+      - name: ${{ inputs.workflow-tag }}
+        if: inputs.workflow-tag != ''
+        run: |
+          # Do nothing
+  # Each section here could be split out into a separate job, at the cost of slightly increased complexity.
+  # This would improve the (already somewhat fast) performance a bit, but I'm not sure if it's worth the
+  # tradeoff.
+  build-plugins:
+    needs:
+      - setup
+    runs-on: ubuntu-22.04-32core
+    environment: ${{ needs.setup.outputs.environment }}
+    permissions:
+      contents: read
+      id-token: write
+    env:
+      ARTIFACT_DIRECTORY: /tmp/build
+    steps:
+      # Setup
+      - name: Enable performance telemetry/metrics
+        uses: catchpoint/workflow-telemetry-action@v1
+        with:
+          comment_on_pr: false
+      - name: Checkout repo
+        uses: actions/checkout@v4
+        with:
+          ref: ${{ needs.setup.outputs.gitref }}
+          fetch-depth: 0  # This is required by some of the commands in the makefiles
+      - name: Setup Go
+        uses: actions/setup-go@v5
+        with:
+          go-version-file: "./go.mod"
+          cache-dependency-path: "./go.sum"
+          check-latest: true
+      - name: Set environment variables for Makefiles
+        env:
+          VERSION_TAG: ${{ needs.setup.outputs.version }}
+        run: |
+          {
+            echo "VERSION=${VERSION_TAG##v}"
+            echo "GITREF=$VERSION_TAG" >> "$GITHUB_ENV"
+            echo "GNUMAKEFLAGS=-j$(nproc)"
```

Comment:
I think this `-j$(nprocs)` hurts more that it helps. The Go compiler is already very concurrent itself. Running multiple at the same time just adds CPU contention as well as go cache contention.

I did a test:
```
$ sudo rm -rf cache; mkdir -p cache/{build,pkg}
$ GOMODCACHE=$(pwd)/cache/pkg GOCACHE=$(pwd)/cache/build time make releases

vs

$ sudo rm -rf cache; mkdir -p cache/{build,pkg}
$ GOMODCACHE=$(pwd)/cache/pkg GOCACHE=$(pwd)/cache/build time make -j$(nproc) releases
```

The former took 1m8s on my 16 core laptop. The latter (`-j$(nproc)` took 2m10s. There were also concurrent downloads of the Go toolchain and many packages that are needed across the builds. For the former, my load average peaked at 16. For the latter, it peaked at 72 and brought my machine to a crawl while running.

That concurrency means the Makefiles need to be written correctly. I'm not sure they are and things that are implicitly ordered due to a single thread of execution may end up in the wrong order producing incorrect builds or inconsistent/flaky results.

But I think the numbers themselves are enough to not use `-j`.

Reply:
> There were also concurrent downloads of the Go toolchain and many packages that are needed across the builds.

I meant to download the dependencies before hand via `go mod download`. I think I just accidentally deleted the line during some testing. The go modules should also be cached by the setup go step, so there should be little contention after the first run. Between these two I don't think that there should be any concurrent downloads of Go modules.

> That concurrency means the Makefiles need to be written correctly. I'm not sure they are and things that are implicitly ordered due to a single thread of execution may end up in the wrong order producing incorrect builds or inconsistent/flaky results.

There were one or two things, but I fixed the issue and they now work in parallel.

If you'd like I can do a couple of test runs in a GHA context with and without the flag set. I think the results here may be different than your local results due to caching, and after ensuring module downloads occur once.

Reply:
My intuition was that `make -jN` was not going to be useful because the Go compiler will make use of all cores anyway, and now requires that people be careful (and know to be careful and know how to be careful) with the Makefiles (not borne out in practice). I was quite surprised that `-jN` took twice as long, but not surprised the load average went through the roof.

The other concern with `-jN` is it interleaves the output and makes the logs largely useless. Overall I think it's a net loss on multiple factors to use `-jN`.

Reply:
> The other concern with -jN is it interleaves the output and makes the logs largely useless

Yea this part is... not great.

After I finish addressing the other comments I'll run a couple of builds with and without `-j`. If there is a significant speedup when running in a GHA context then I'll consider leaving the flag, otherwise I'll remove it. At the very least I will split a couple of the `make` invocations up so that logs are grouped a little better (for example, I'll split up the `binaries` and `helm-charts` target invocations).

Reply:
I did some testing and it looks like there is an improvement, but it's not huge. The largest improvement looks to be during the image creation portion, which basically does the entire build again. Note that this test was done with a miniscule sample size, and the data I collected may not be indicative of the population as a whole.

I'm going to remove the flag. While it does improve performance a bit in some areas, I don't think that the tradeoff is worth it.